concordium-contracts-common-derive/src/lib.rs

```diff
@@ -155,9 +155,75 @@ pub fn deserial_with_state_derive(input: TokenStream) -> TokenStream {
     unwrap_or_report(derive::impl_deserial_with_state(&ast))
 }
 
-/// Derive the `SchemaType` trait for a type.
-/// If the feature `build-schema` is not enabled this is a no-op, i.e., it does
-/// not produce any code.
+/// Derive the [`SchemaType`] trait for a type matching the implementation when
+/// deriving [`Serial`].
+///
+/// Can be used for enums and structs.
+/// If the type is a struct all fields must implement the [`SchemaType`] trait.
+/// If the type is an enum then all fields of each of the enums must implement
+/// the [`SchemaType`] trait.
+///
+/// ## Generic type bounds
+///
+/// By default a trait bound is added on each generic type for implementing
+/// [`SchemaType`]. However, if this is not desirable, the bound can be put
+/// explicitly using the `bound` attribute on the type overriding the default
+/// behavior by adding the provided bound for the implementation.
+///
+/// Bounds present in the type declaration will still be present in
+/// the implementation, even when a bound is provided:
+///
+/// ### Example
+///
+/// ```ignore
+/// #[derive(SchemaType)]
+/// #[concordium(bound(schema_type = "A: SomeOtherTrait"))]
+/// struct Foo<A: SomeTrait> {
+///     bar: A,
+/// }
+///
+/// // Derived implementation:
+/// impl <A: SomeTrait> SchemaType for Foo<A> where A: SomeOtherTrait { .. }
+/// ```
+///
+/// ## Collections
+///
+/// Collections (Vec, BTreeMap, BTreeSet) and strings (String, str) can be
+/// annotated with `size_length` which is the number of bytes used for encoding
+/// the number of elements, see derive macro ['Serial'] for more on this.
+///
+/// The value of this field is the number of bytes that will be used for
+/// encoding the number of elements. Supported values are `1`, `2`, `4`, `8`.
+///
+/// ### Example
+/// ```ignore
+/// #[derive(SchemaType)]
+/// struct Foo {
+///     #[concordium(size_length = 1)]
+///     bar: BTreeSet<u8>,
+/// }
+/// ```
+///
+/// ## Transparent
+///
+/// Deriving [`SchemaType`] for structs using the newtype design pattern exposes
+/// the wrapping struct which is often not desirable. The attribute
+/// `#[concordium(transparent)]` can be added above the struct which changes the
+/// implementation of [`SchemaType`] to schema type of the inner field.
+///
+/// The attribute can `#[concordium(transparent)]` attribute can only be used
+/// for structs with a single field, and the type of this field must implement
+/// `SchemaType`.
```

Comment:
There is a bit more complexity here. In a struct

```
struct Foo {
   #[concordium(size_length = 4)]
   inner: Vec<SomeType>
}
```

The field attribute is still used. This is important as well, but it is not obvious that this is a choice that was made.

So can you please also add an example with a vector or some other container and explain the semantics.

Reply:
I have added a bit more to this